### ISSUE TITLE
utcbnldigeetichf

### ISSUE THREAD

**Issue by jlegrone** [cnabio/cnab-spec#337]
# Background

#285 has raised important points about the need to avoid granting overly permissive authority to invocation images via CNAB credentials.

The spec does not currently describe a mechanism for allowing bundles to communicate to a runtime what permissions they require in order to deploy to the underlying platform target. This problem is not specific to Kubernetes, but because Kubernetes deployments are a major use case for early adopters of CNAB, this is a useful context for discussing possible solutions.

# Proposal

## Goal

The goal is to add a non-normative extension to the spec which
- allows runtimes to authorize Kubernetes API scopes requested by bundles
- is agnostic to the underlying application definition, rendering, or deployment mechanism

### 1. Enable bundles to specify required Kubernetes API permissions

Bundles that deploy to Kubernetes SHOULD implement the Kubernetes RBAC [custom extension](https://github.com/cnabio/cnab-spec/blob/master/101-bundle-json.md#custom-extensions):

```json
{
    "custom": {
        "io.kubernetes.rbac": {
            "version": "v1alpha1",
            "rules": [
                {
                    "apiGroups": [""],
                    "resources": ["pods", "pods/log"],
                    "verbs": ["get", "watch", "list"]
                },
                {
                    "apiGroups": ["extensions", "apps"],
                    "resources": ["deployments"],
                    "verbs": ["get", "watch", "create", "update", "patch", "delete"]
                }
            ]
        }
    }
}
```

### 2. Allow runtimes to identify which CNAB credentials are Kubernetes credentials

The spec already describes several [well-known custom actions](https://github.com/cnabio/cnab-spec/blob/master/804-well-known-custom-actions.md#well-known-custom-actions), which are just custom actions with scoped names.

Bundles that deploy Kubernetes applications would then declare a credential named `io.k8s.kubeconfig`. CNAB runtimes would be free to construct a kubeconfig

**Comment by technosophos** [comment]
My understanding is that you are declaring a new custom extension that would allow embedding k8s RBACs, right? It's a little confusing when you way:

> We can follow a similar pattern to that laid out by well-known custom actions to use well-known CNAB parameter and credential names, along with a custom extension for configuring RBAC, to achieve these objectives.

But the data structure you use looks like a custom extension.

That sounds like a great idea to me. As one of the non-required extensions, this would make it easy for Kubernetes users to make use of this without impacting the non-Kubernetes cases.

I think it could also give us a reasonable pattern to suggest for other agents that want to embed authz data.

**Comment by jlegrone** [comment]
> But the data structure you use looks like a custom extension.

Yes, this issue describes a custom extension. So far I haven't seen any extensions that also leverage naming conventions for credentials or parameters, while this proposal does both. I referenced "well known" custom action names to demonstrate that this is an established pattern for other areas of the spec.

**Comment by technosophos** [comment]
Excellent! I am in favor of this proposal.

**Comment by carolynvs** [comment]
I asked a few questions about this at last week's meeting and wanted to copy some of the answers here (from the meeting notes) just so that people can see them:

- Carolyn: Is idea that bundle says these are the credentials the bundle needs and that the expectation is that the runtime creates user/service config with these?
  - Jacob: Idea is to pass admin credentials and have them down privileged. Unfortunately this isn't practically easy to do
- Carolyn: Where would the code that does this live, CNAB Go?
  - Matt: This would be non-normative so could be implemented in an external tool
  - Jacob: For CNAB Go, we could have an interface for multiple credential providers and that spits out credential sets
  - Carolyn: So this would be performed before the bundle is executed so the injected credential is already transformed.
- Carolyn: Wondering how portable bundles like this would be
  - Jacob: Goal is to have something portable and avoid having bundles only be able to be installed by one tool
  - Carolyn: Tool that understands this extension has better security, naive tool still works
